cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -411,40 +411,104 @@ void TransformMatchSubstring(const uint8_t* pattern, int64_t pattern_length,
+                                         output_offset, output);
+  }
+};
+
+template <typename offset_type>
+struct RegexSubstringMatcher {
+  const MatchSubstringOptions& options_;
+  const RE2 regex_match_;
+
+  RegexSubstringMatcher(KernelContext* ctx, const MatchSubstringOptions& options)
+      : options_(options), regex_match_(options_.pattern) {
+    if (!regex_match_.ok()) {
+      ctx->SetStatus(Status::Invalid("Regular expression error"));
+    }
+  }
+
+  void Match(const offset_type* offsets, const uint8_t* data, int64_t length,
+             int64_t output_offset, uint8_t* output) {
+    FirstTimeBitmapWriter bitmap_writer(output, output_offset, length);
+    for (int64_t i = 0; i < length; ++i) {
+      const char* current_data = reinterpret_cast<const char*>(data + offsets[i]);
+      int64_t current_length = offsets[i + 1] - offsets[i];
+      auto piece = re2::StringPiece(current_data, current_length);
+      if (re2::RE2::PartialMatch(piece, regex_match_)) {
+        bitmap_writer.Set();
+      }
+      bitmap_writer.Next();
+    }
+    bitmap_writer.Finish();
+  }
+};
+
 const FunctionDoc match_substring_doc(
     "Match strings against literal pattern",
     ("For each string in `strings`, emit true iff it contains a given pattern.\n"
      "Null inputs emit null.  The pattern must be given in MatchSubstringOptions."),
     {"strings"}, "MatchSubstringOptions");
 
+const FunctionDoc match_substring_regex_doc(
+    "Match strings against regex pattern",
+    ("For each string in `strings`, emit true iff it matches a given pattern at any "
+     "position.\n"
+     "Null inputs emit null.  The pattern must be given in MatchSubstringOptions."),
+    {"strings"}, "MatchSubstringOptions");
+
 void AddMatchSubstring(FunctionRegistry* registry) {
-  auto func = std::make_shared<ScalarFunction>("match_substring", Arity::Unary(),
-                                               &match_substring_doc);
-  auto exec_32 = MatchSubstring<StringType>::Exec;
-  auto exec_64 = MatchSubstring<LargeStringType>::Exec;
-  DCHECK_OK(func->AddKernel({utf8()}, boolean(), exec_32, MatchSubstringState::Init));
-  DCHECK_OK(
-      func->AddKernel({large_utf8()}, boolean(), exec_64, MatchSubstringState::Init));
-  DCHECK_OK(registry->AddFunction(std::move(func)));
+  {
+    auto func = std::make_shared<ScalarFunction>("match_substring", Arity::Unary(),
+                                                 &match_substring_doc);
+    auto exec_32 = MatchSubstring<StringType, PlainSubstringMatcher>::Exec;
+    auto exec_64 = MatchSubstring<LargeStringType, PlainSubstringMatcher>::Exec;
+    DCHECK_OK(func->AddKernel({utf8()}, boolean(), exec_32, MatchSubstringState::Init));
+    DCHECK_OK(
+        func->AddKernel({large_utf8()}, boolean(), exec_64, MatchSubstringState::Init));
+    DCHECK_OK(registry->AddFunction(std::move(func)));
+  }
+#ifdef ARROW_WITH_RE2
+  {
+    auto func = std::make_shared<ScalarFunction>("match_substring_regex", Arity::Unary(),
```

Comment:
New functions need be documented in https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst

@pitrou , should we also document new functions in python api document?
https://github.com/apache/arrow/blob/master/docs/source/python/api/compute.rst

Reply:
Ah, good point! We should indeed add the names there to auto-generate the API docs. I keep forgetting about it, so there's probably more to add.